File: android/src/main/java/com/elylucas/capscreenbrightness/ScreenBrightnessPlugin.kt
package com.elylucas.capscreenbrightness

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.PluginThread
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "ScreenBrightness")
public class ScreenBrightnessPlugin : Plugin() {
    // The window's attributes belong to the main thread
    @PluginMethod(thread = PluginThread.MAIN)
    public fun setBrightness(call: PluginCall) {
        val brightness = call.getFloat("brightness")
        val layoutParams = activity.window.attributes

        // A call without a brightness throws here, as it always has; on the MAIN thread the bridge rejects the call
        // with what it throws instead of letting it crash the app.
        layoutParams.screenBrightness = brightness!!
        activity.window.attributes = layoutParams
        call.resolve()
    }

    @PluginMethod
    public fun getBrightness(call: PluginCall) {
        val layoutParams = activity.window.attributes
        val ret = JSObject()
        // Written as a double, which is what Java widened the float to.
        ret.put("brightness", layoutParams.screenBrightness.toDouble())
        call.resolve(ret)
    }
}

File: ios/Sources/ScreenBrightnessPlugin/ScreenBrightnessPlugin.swift
import Foundation
import UIKit
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ScreenBrightnessPlugin)
public class ScreenBrightnessPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ScreenBrightnessPlugin"
    public let jsName = "ScreenBrightness"
    public let pluginMethods: [CAPPluginMethod] = [
        .promise("setBrightness", ScreenBrightnessPlugin.setBrightness),
        .async("getBrightness", ScreenBrightnessPlugin.getBrightness)
    ]

    /// The setter stays synchronous: the bridge queue runs it in the order of the calls and it hands its UIKit work to
    /// the main queue in that order, so the last call wins. An async method would not keep that order.
    func setBrightness(_ call: CAPPluginCall) {
        let brightness = call.getFloat("brightness")
        DispatchQueue.main.async {
            // Without a brightness, the current one: UIScreen is read on the main thread, like it is written.
            UIScreen.main.brightness = CGFloat(brightness ?? Float(UIScreen.main.brightness))
            call.resolve()
        }
    }

    /// UIScreen belongs to the main thread, so the read runs on the main actor instead of the bridge queue. CGFloat is a
    /// Double, which JavaScript receives as the same number.
    @MainActor
    func getBrightness(_ call: CAPPluginCall) async -> JSObject {
        return ["brightness": Double(UIScreen.main.brightness)]
    }
}
